porcupine/actions.py

```diff
@@ -0,0 +1,140 @@
+
+
+@dataclass(kw_only=True, slots=True, frozen=True)
+class PathAction:
+    """Action that requires a Path to be provided to the callback"""
+
+    name: str
+    description: str
+    callback: Callable[[Path], None]
+    file_compatible: bool = True
+    directory_compatible: bool = False
+    availability_callbacks: list[path_action_availability_callback] | None = None
+
+
+ActionTypes = Action | FileTabAction | PathAction
+
+_actions: dict[str, ActionTypes] = {}
+
+
+def register_action(
+    *,
+    name: str,
+    description: str,
+    callback: Callable[..., None],
+    availability_callbacks: list[action_availability_callback] | None = None,
+) -> Action:
+    if name in _actions:
+        raise ValueError(f"Action with the name '{name}' already exists")
+    action = Action(
+        name=name,
+        description=description,
+        callback=callback,
+        availability_callbacks=availability_callbacks,
+    )
+    _actions[name] = action
+    return action
+
+
+def register_filetab_action(
+    *,
+    name: str,
+    description: str,
+    callback: Callable[[FileTab], None],
+    availability_callbacks: list[filetab_action_availability_callback] | None,
+) -> FileTabAction:
+    if name in _actions:
+        raise ValueError(f"Action with the name '{name}' already exists")
+    action = FileTabAction(
+        name=name,
+        description=description,
+        callback=callback,
+        availability_callbacks=availability_callbacks,
+    )
+    _actions[name] = action
+    return action
+
+
+def register_path_action(
+    *,
+    name: str,
+    description: str,
+    callback: Callable[[Path], None],
+    file_compatible: bool = True,
+    directory_compatible: bool = False,
+    availability_callbacks: list[path_action_availability_callback] | None,
+) -> PathAction:
+    if name in _actions:
+        raise ValueError(f"Action with the name '{name}' already exists")
+    action = PathAction(
+        name=name,
+        description=description,
+        callback=callback,
+        file_compatible=file_compatible,
+        directory_compatible=directory_compatible,
+        availability_callbacks=availability_callbacks,
+    )
+    _actions[name] = action
+    return action
+
+
+def filetype_availability(filetypes: list[str]) -> Callable[[FileTab | Path], bool]:
+    def _filetype_availability(filetypes: list[str], context: FileTab | Path) -> bool:
+        if isinstance(context, FileTab):
+            tab = context
+            if tab.settings.get("filetype_name", object) in filetypes:
+                return True
+            return False
+
+        if isinstance(context, Path):
+            path = context
+
+            if not path.exists():
+                raise RuntimeError(f"{path} does not exist.")
+            if path.is_dir():
+                raise RuntimeError(
+                    f"{path} is a directory - an action consumer registered this action incorrectly"
```

Comment:
IMO it should just return `False`. The command palette has everything "registered" anyways, and it will need to test the availability of all actions when finding the one that the user wants.

Reply:
I need to think about this more, but the thought I had is that passing a directory to this availability function doesn't make sense; directories don't have filetypes.

Maybe we need a  4th action type: one that takes directories.

Reply:
I imagined it would be a boolean in `PathAction`. So if you set the flag to `True`, it makes the action available for directories without even calling the user's availability callback.

Reply:
Well, see that gets tricky. Some actions might **only** work on a directory, some might work on both files and directories.

Also there is the question of recursion, not exactly sure how that should be handled :sweat:

Reply:
Let's make it two booleans: does it work with files (default: True), and does it work with directories (default: False). Now any combination is possible, and all except False+False make sense, so maybe throw an error if both are False.

We should solve recursion later, but my hunch is that the best solution is to do nothing, since searching the file system recursively can be slow (e.g. accidentally right-clicking `/home` on a computer that doesn't have an SSD).

Reply:
To make this easy for consumers, I'd suggest adding a new `is_available(self, path: Path)` method that considers the booleans, and if needed, invokes the callback.

To be clear, the booleans will be a convenience thing that allows us to skip calling the callback when e.g. the path is a directory but we expect a file. You could just check it in the callback, but it seems easy to forget to return `False` for directories, so booleans could help with that.

Now I'm no longer sure if we need two booleans. If you only want directories, set `accepts_directories=True` and make your callback function reject non-directories?

Reply:
It feels like we have basically come full circle :sweat_smile:  why not just have availability callbacks for `accepts_directories` and `accepts_files`?

Reply:
Indeed :D

If we go with my `lambda` idea from the other comment, we don't even need to predefine availability callbacks for this: path objects already have convenient `.is_file()` and `.is_dir()` methods :)

Reply:
Yep! Sounds good!

Reply:
tracked in https://github.com/Akuli/porcupine/issues/1357